File: Auxiliary_tools/Read_muSTEM_output.m
function [ array_out ] = read_muSTEM_output( title,fmt )
% The function binary_in(title,fmt) reads in a binary file outputed from the MuSTEM
% software. The variable keyword should be a string containing the filepath. The variable
% fmt is optional and specifies the endianness of the file. Use 'b' for Big-endian (default) and 'l'
% for Little-endian
    %Get filename as seperate from directory
	[~,name,~] = fileparts(title);
     
	%Assert the existence of the file
    assert(exist(title,'file')~=1,strcat('Could not find file ',title));
	
	%Pass filename for array dimensions
    dimensions = regexp(name,'([0-9]+)x([0-9]+)','tokens');
    dimensions = dimensions{1};
    Npx = str2num(dimensions{1});
    Npy = str2num(dimensions{2});
   
    %With knowledge of the array dimensions, the file datatype can be intuited   
    s = dir(title);
    type = s.bytes/Npx/Npy;
    if(type==8)
        format='double';
    else
        format='single';
    end
   
    %Get endianness
    if (nargin>1)
        fmt_= fmt;
    else
        fmt_ = 'b';
    end
   
    %Read array from binary file
    fileID = fopen(title);
    array_out = fread(fileID,[Npx Npy],format,0,fmt_);
    fclose(fileID);
end

File: Auxiliary_tools/Read_muSTEM_output.py
import re,os,numpy as np

def open_muSTEM_binary(filename):
	'''opens binary with name filename outputted from the muSTEM software'''
	assert os.path.exists(filename), filename+' does not exist'
	#Parse filename for array dimensions
	m = re.search('([0-9]+)x([0-9]+)',filename)
	if m:
		y = int(m.group(2))
		x = int(m.group(1))
	#Get file size and intuit datatype
	size =  os.path.getsize(filename)
	if (size/(y*x) == 4):
		dtype = '>f4'
	elif(size/(y*x) == 8):
		dtype = '>f8'
	print dtype
	#Read data and reshape as required.
	return np.reshape(np.fromfile(filename,dtype = dtype),(y,x))
